README.md
# UtiliKit

![CI Status](https://github.com/BottleRocketStudios/iOS-UtiliKit/actions/workflows/main.yml/badge.svg)
[![Version](https://img.shields.io/cocoapods/v/UtiliKit.svg?style=flat)](http://cocoapods.org/pods/UtiliKit)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage)
[![SwiftPM compatible](https://img.shields.io/badge/SwiftPM-compatible-4BC51D.svg?style=flat)](https://swift.org/package-manager)
[![Platform](https://img.shields.io/cocoapods/p/UtiliKit.svg?style=flat)](http://cocoapods.org/pods/UtiliKit)
[![codecov](https://codecov.io/gh/BottleRocketStudios/iOS-UtiliKit/branch/master/graph/badge.svg)](https://codecov.io/gh/BottleRocketStudios/iOS-UtiliKit)
[![License](https://img.shields.io/cocoapods/l/UtiliKit.svg?style=flat)](http://cocoapods.org/pods/UtiliKit)

**UtiliKit** provides several useful and often common additions for iOS applications. These extensions, protocols, and structs are designed to simplify boilerplate code as well as remove common "String-typed" use cases.

## Contents

The following implementations are provided:

* [**ActiveLabel**](Documentation/READMEs/ActiveLabel.md) - This subspec provides a `UILabel` subclass that renders gradient "loading" animations while the label's `text` property is set to `nil`.
* [**Container**](Documentation/READMEs/Container.md) - This subspec provides a simple `ContainerViewController` without any built-in navigation construct.
* [**General**](Documentation/READMEs/General.md) - This subspec includes extensions for both `FileManager` and `UIView`. These simplify getting common URLs and programmatically adding views down to simple variables and function calls.
* [**Instantiation**](Documentation/READMEs/Instantiation.md) - This subspec changes "String-typed" view instantiation, view controller instantiation, and reusable view dequeuing into type-safe function calls.
* [**Obfuscation**](Documentation/READMEs/Obfuscation.md) - This subspec provides simple routines to remove plaintext passwords or keys from your source code.
* [**ScrollingPageControl**](Documentation/READMEs/ScrollingPageControl.md) - This subspec provides a view modeled from Apple's `UIPageControl` to allow representation of a large number of pages.
* [**TimelessDate**](Documentation/READMEs/TimelessDate.md) - This subspec is an abstraction away from `Date` and `Calendar`. It is primarily designed to be used for simple scheduling and day comparisons in which the time is less important that the actual day.
* [**Version**](Documentation/READMEs/Version.md) - This subspec simplifies the display of version and build numbers.

## Example

Clone the repo:

```bash
git clone https://github.com/BottleRocketStudios/iOS-UtiliKit.git
```

From here, you can open up `UtiliKit.xcworkspace` and run the "UtiliKit-iOSExample" project.

## Requirements

* iOS 10.0+
* Swift 5.1

## Installation

### Cocoapods

UtiliKit is available through [CocoaPods](http://cocoapods.org). To install it, simply add the following line to your Podfile:

```ruby
pod 'UtiliKit'
```

### Carthage

Add the following to your [Cartfile](https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#cartfile):

```
github "BottleRocketStudios/iOS-UtiliKit"
```

Run `carthage update` and follow the steps as described in Carthage's [README](https://github.com/Carthage/Carthage#adding-frameworks-to-an-application).

### Swift Package Manager

```swift
dependencies: [
    .package(url: "https://github.com/BottleRocketStudios/iOS-UtiliKit.git", from: "1.8.0")
]
```

You will then need to choose from the available libraries to add to your project. These libraries should match up with the subspecs available through Cocoapods.

* `UtiliKit` - Imports all of the below libraries available individually.
* `ActiveLabel` - This subspec provides a `UILabel` subclass that renders gradient "loading" animations while the label's `text` property is set to `nil`.
* `ContainerViewController` - This subspec provides a simple `ContainerViewController` without any built-in navigation construct.
* `GeneralUtilities` - This subspec includes extensions for both `FileManager` and `UIView`. These simplify getting common URLs and programmatically adding views down to simple variables and function calls.
* `Instantiation` - This subspec changes "Stringly-typed" view instantiation, view controller instantiation, and reusable view dequeuing into type-safe function calls.
* `Obfuscation` - This subspec provides simple routines to remove plaintext passwords or keys from your source code.
* `ScrollingPageControl` - This subspec provides a view modeled from Apple's `UIPageControl` to allow representation of a large number of pages.
* `TimelessDate` - This subspec is an abstraction away from `Date` and `Calendar`. It is primarily designed to be used for simple scheduling and day comparisons in which the time is less important that the actual day.
* `Versioning` - This subspec simplifies the display of version and build numbers.

## Author

[Bottle Rocket Studios](https://www.bottlerocketstudios.com/)

## License

UtiliKit is available under the Apache 2.0 license. See the LICENSE.txt file for more info.

## Contributing

See the [CONTRIBUTING] document. Thank you, [contributors]!

[CONTRIBUTING]: CONTRIBUTING.md
[contributors]: https://github.com/BottleRocketStudios/iOS-UtiliKit/graphs/contributors


CHANGELOG.md
## Main

##### Enhancements

* None

##### Bug Fixes

* Fix debug symbol not showing in `verboseVersionString` method.
[Earl Gaspard](https://github.com/earlgaspard)
[#121](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/121)


## 1.8.0 (2022-01-19)

##### Enhancements

* Extend CLLocationCoordinate2D with isValid computed property
[Andrew Winn](https://github.com/andrew-winn-br)
[#102](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/102)

* Improve support for registering supplementary and decoration views with UICollectionView.
[Will McGinty](https://github.com/willmcginty)
[#110](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/110)

* Re-organize the way Swift Packages are managed. Instead of importing `UtiliKit` as a package, each target (or Cocoapods subspec) is imported separately. The full list of libraries is: `GeneralUtilities`, `Instantiation`, `TimelessDate`, `Versioning`, `ContainerViewController`, `ActiveLabel`, `Obfuscation`. `UtiliKit` remains as a possible library which will import all the above targets at once.
[Will McGinty](https://github.com/willmcginty)
[#104](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/104)

##### Bug Fixes

* None


## 1.7.0 (2020-08-17)

##### Enhancements

* Make it possible to override functions in container, which aid in `ManagedChild` management.
[Will McGinty](https://github.com/willmcginty)
[#87](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/87)

* Add `postTransitionBehavior` to `ContainerViewController`, automating some common child management scenarios
[Dimitar Milinski](https://github.com/dmilinski)
[#88](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/88)

* Switch `removeAllNonVisibleChildren` to `public`
[Dimitar Milinski](https://github.com/dmilinski)
[#90](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/90)

* Add `ScrollingPageControl`
[Nathan Chiu](https://github.com/BR-Nathan-Chiu)
[#95](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/95)

##### Bug Fixes

* None


## 1.6.1 (2019-10-04)

##### Enhancements

* None

##### Bug Fixes

* Fix an issue where the completion handler does not get called when transitioning.
[Will McGinty](https://github.com/willmcginty)
[#76](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/76)

* Fix an issue where the transition was mistakenly marked as a failure.
[Will McGinty](https://github.com/willmcginty)
[#77](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/77)

* Fix a transitionting issue where appearance callbacks were unbalanced in iOS 13.
[Will McGinty](https://github.com/willmcginty)
[#79](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/79)

* Fix an issue where ContainerViewController was not removing source or destination views from the view hierarchy after transitioning
[Dimitar Milinski](https://github.com/dmilinski)
[#81](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/81)


## 1.6.0 (2019-08-29)

##### Enhancements

* Added Swift Package Manager support.
[Brian Miller](https://github.com/jobsismyhomeboy)
[#64](https://github.com/BottleRocketStudios/iOS-UtiliKit/issues/64)

* Add interactive transitioning and transition coordination to `ContainerViewController`. This change has bumped the deployment target to iOS 10.0+
[Will McGinty](https://github.com/willmcginty)
[#63](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/63)

* Added the `Obfuscation` subspec, which includes the `ObfuscatedKey` structure for keys/passwords to ensure that
they don't appear in plaintext within the source or binary of your app.
[Russell Mirabelli](https://github.com/rmirabelli)
[#67](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/67)

* Updated initial/named view controller example so that the buttons are horizontally centered in the stack view.
[Tyler Milner](https://github.com/tylermilner)
[#65](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/65)

* Conform URL to ExpressibleByStringLiteral (e.g. `let url: URL = "www.apple.com"`)
[Will McGinty](https://github.com/willmcginty)
[#66](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/66)

* Added Carthage support.
[Ryan Gant](https://github.com/ganttastic)
[#68](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/68)

* Updated `README` to include the "ActiveLabel" subspec listed at the top and added code examples for the "General" subspec.
[Tyler Milner](https://github.com/tylermilner)
[#71](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/71)

##### Bug Fixes

* None


## 1.5.0 (2019-07-11)

##### Enhancements

* Add `ActiveLabel` class to help show activity on a label.
[Brian Miller](https://github.com/JobsIsMyHomeboy)
[#55](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/55)

* Add support for `Configurable` types when dequeuing reusable views.
[Will McGinty](https://github.com/willmcginty)
[#53](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/58)

##### Bug Fixes

* Fix a bug where completion wasn't always called when transitioning.
[Will McGinty](https://github.com/willmcginty)
[#57](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/57)


## 1.4.0 (2019-04-30)

##### Enhancements

* Migrate to Swift 5.0.
[Earl Gaspard](https://github.com/earlgaspard)
[#53](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/53)

##### Bug Fixes

* Fix an issue where right and bottom insets were being inverted in constraints.
[Will McGinty](https://github.com/willmcginty)
[#46](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/50)


## 1.3.5 (2019-01-02)

##### Enhancements

* None.

##### Bug Fixes

* Handle another case of transitioning pre-load.
[Will McGinty](https://github.com/willmcginty)
[#46](https://github.com/BottleRocketStudios/iOS-UtiliKit/pull/46)


## 1.3.4 (2018-12-19)

##### Enhancements

* Make interface of `UICollectionView.SupplementaryElementKind` public.
[Will McGinty](https://github.com/willmcginty)

Package.swift
// swift-tools-version:5.1
import PackageDescription

let package = Package(
    name: "UtiliKit",
    platforms: [
        .iOS("10.0")
    ],
    products: [
        .library(
            name: "UtiliKit",
            targets: ["GeneralUtilities", "Instantiation", "TimelessDate", "Versioning",
                      "ContainerViewController", "ActiveLabel", "Obfuscation", "ScrollingPageControl"]),
        .library(name: "GeneralUtilities", targets: ["GeneralUtilities"]),
        .library(name: "Instantiation", targets: ["Instantiation"]),
        .library(name: "TimelessDate", targets: ["TimelessDate"]),
        .library(name: "Versioning", targets: ["Versioning"]),
        .library(name: "ContainerViewController", targets: ["ContainerViewController"]),
        .library(name: "ActiveLabel", targets: ["ActiveLabel"]),
        .library(name: "Obfuscation", targets: ["Obfuscation"]),
        .library(name: "ScrollingPageControl", targets: ["ScrollingPageControl"])
        ],
    targets: [
        .target(name: "GeneralUtilities", path: "Sources/General"),
        .target(name: "Instantiation", path: "Sources/Instantiation"),
        .target(name: "TimelessDate", path: "Sources/TimelessDate"),
        .target(name: "Versioning", path: "Sources/Version"),
        .target(name: "ContainerViewController", path: "Sources/Container"),
        .target(name: "ActiveLabel", path: "Sources/ActiveLabel"),
        .target(name: "Obfuscation", path: "Sources/Obfuscation"),
        .target(name: "ScrollingPageControl", dependencies: [.target(name: "GeneralUtilities")], path: "Sources/ScrollingPageControl")
    ]
)

